api/rosetta/errors.go

```diff
@@ -33,105 +31,129 @@ type Error struct {
 	Details     map[string]interface{} `json:"details,omitempty"`
 }
 
-func Internal(err error) Error {
-	return Error{
-		ErrorDefinition: configuration.ErrorInternal,
-		Description:     err.Error(),
-		Details:         nil,
+type DetailFunc func(map[string]interface{})
+
+func WithError(err error) DetailFunc {
+	return func(details map[string]interface{}) {
+		details["error"] = err.Error()
 	}
 }
 
-func InvalidFormat(message string, args ...interface{}) Error {
-	return Error{
-		ErrorDefinition: configuration.ErrorInvalidFormat,
-		Description:     fmt.Sprintf(message, args...),
-		Details:         nil,
+func WithDetail(key string, val interface{}) DetailFunc {
+	return func(details map[string]interface{}) {
+		details[key] = val
 	}
 }
 
-func InvalidNetwork(fail failure.InvalidNetwork) Error {
-	return Error{
-		ErrorDefinition: configuration.ErrorInvalidNetwork,
-		Description:     fail.Message,
-		Details: map[string]interface{}{
-			"blockchain": fail.Blockchain,
-			"network":    fail.Network,
-		},
+func RosettaError(definition meta.ErrorDefinition, description string, details ...DetailFunc) Error {
+	dd := make(map[string]interface{})
+	for _, detail := range details {
+		detail(dd)
 	}
+	e := Error{
+		ErrorDefinition: definition,
+		Description:     description,
+		Details:         dd,
+	}
+	return e
+}
+
+func Internal(description string, err error) Error {
+	return RosettaError(
+		configuration.ErrorInternal,
+		description,
+		WithError(err),
+	)
+}
+
+func InvalidEncoding(description string, err error) Error {
+	return RosettaError(
+		configuration.ErrorInvalidEncoding,
+		description,
+		WithError(err),
+	)
+}
+
+func InvalidFormat(description string, details ...DetailFunc) Error {
+	return RosettaError(
+		configuration.ErrorInvalidFormat,
+		description,
+		details...,
+	)
+}
+
+func ConvertError(definition meta.ErrorDefinition, description failure.Description, details ...DetailFunc) Error {
```

Comment:
This could also be unexported. Plus this might be more of an error _creation_ than error _conversion_.

Reply:
Yes, not exporting it makes sense. However, I think conversion does make more sense than creation, personally. Creating an error is from scratch, conversion is creating it from something else. In our case, that's actually the error that is taken apart and then forwarded to the function.